host/class/uac/usb_host_uac/uac_host.c

```diff
@@ -0,0 +1,2141 @@
+        UAC_GOTO_ON_ERROR(_uac_host_device_add(config->addr, dev_hdl, config_desc, &uac_device), "Unable to add UAC device");
+        new_device = true;
+        uac_device->opened_cnt = 0;
+    }
+
+    UAC_GOTO_ON_ERROR(uac_host_interface_add(uac_device, config->iface_num, &uac_iface), "Unable to add interface");
+
+    // Save UAC Interface callback
+    uac_iface->user_cb = config->callback;
+    uac_iface->user_cb_arg = config->callback_arg;
+    // create a ringbuffer for the incoming/outgoing data
+    uac_iface->ringbuf = xRingbufferCreate(config->buffer_size, RINGBUF_TYPE_BYTEBUF);
+    UAC_GOTO_ON_FALSE(uac_iface->ringbuf, ESP_ERR_NO_MEM, "Unable to create ringbuffer");
+    uac_iface->ringbuf_size = config->buffer_size;
+    // if the threshold is not set, set it to 25% of the buffer size
+    uac_iface->ringbuf_threshold = config->buffer_threshold ? config->buffer_threshold : config->buffer_size / 4;
+    uac_iface->state = UAC_INTERFACE_STATE_IDLE;
+    *uac_dev_handle = (uac_host_device_handle_t)uac_iface;
+    UAC_ENTER_CRITICAL();
+    uac_device->opened_cnt++;
+    UAC_EXIT_CRITICAL();
+    return ESP_OK;
+
+fail:
+    if (uac_iface) {
+        uac_host_interface_delete(uac_iface);
+    }
+    if (new_device) {
+        _uac_host_device_delete(uac_device);
+    }
+    if (dev_hdl) {
+        usb_host_device_close(s_uac_driver->client_handle, dev_hdl);
+    }
+    if (uac_iface->ringbuf) {
+        vRingbufferDelete(uac_iface->ringbuf);
+    }
+    return ret;
+}
+
+esp_err_t uac_host_device_close(uac_host_device_handle_t uac_dev_handle)
+{
+    uac_iface_t *uac_iface = get_iface_by_handle(uac_dev_handle);
+    UAC_RETURN_ON_INVALID_ARG(uac_iface);
+
+    esp_err_t ret = ESP_OK;
+    ESP_LOGD(TAG, "Close addr %d, iface %d, state %d", uac_iface->dev_info.addr, uac_iface->dev_info.iface_num, uac_iface->state);
+
+    UAC_RETURN_ON_ERROR(uac_host_interface_try_lock(uac_iface, DEFAULT_CTRL_XFER_TIMEOUT_MS), "UAC Interface is busy by other task");
+    if (UAC_INTERFACE_STATE_ACTIVE == uac_iface->state) {
+        UAC_GOTO_ON_ERROR(uac_host_interface_suspend(uac_iface), "Unable to disable UAC Interface");
+    }
+
+    if (UAC_INTERFACE_STATE_READY == uac_iface->state) {
+        UAC_GOTO_ON_ERROR(uac_host_interface_release_and_free_transfer(uac_iface), "Unable to release UAC Interface");
+    }
+
+    // Wait for user delete
+    if (FLAG_INTERFACE_WAIT_USER_DELETE & uac_iface->flags) {
+        uac_host_interface_unlock(uac_iface);
+        return ESP_OK;
+    }
+
+    UAC_ENTER_CRITICAL();
+    if (--uac_iface->parent->opened_cnt == 0) {
+        UAC_EXIT_CRITICAL();
+        UAC_GOTO_ON_ERROR(usb_host_device_close(s_uac_driver->client_handle, uac_iface->parent->dev_hdl), "Unable to close USB device");
+        ESP_LOGD(TAG, "line %d, Close Device addr %d", __LINE__, uac_iface->parent->addr);
+        UAC_GOTO_ON_ERROR(_uac_host_device_delete(uac_iface->parent), "Unable to delete UAC device");
```

Comment:
It seems, that we have: 
1. TAILQ of devices (which is USB device with specific dev_hdl). This is a list of all USB devices, currently connected to the esp. That means, that when device is in this list, it physically attached to the port. 
2. TAILQ of interfaces (which is a uac device, and each of uac has a parent from 1.). 

When use closes the uac device, and it was the only one opened, seems that we don't need to remove the parent. Because physically, the device can be still connected to the port. 
So, it means, that if the we called: `uac_open_dev` - > `uac_close_dev` ->  `uac_open_dev`, w need tp physically reconnect the USB device, because during the third `uac_open_dev` there will be no parent in the list. 

Is that intentional?

Reply:
Hi @roma-jam as the [workflow](https://github.com/espressif/esp-usb/pull/27#issue-2202595629) described.

1. TAILQ of devices, is a list of Opened physical device.
2. TAILQ of interfaces, is a list of Opened logic UAC device

When Open logic UAC device, it will first check if physical device in the list, if not, add the device to the list.

If all logic UAC device closed, the physical device will be removed from list..